src/main.c

```diff
@@ -29,22 +29,53 @@
 #include "exception.h"
 #include "gdb_packet.h"
 #include "morse.h"
+#ifdef ENABLE_RTT
+#include "rtt.h"
+#endif
+
+#define BUF_SIZE 1024U
+static char pbuf[BUF_SIZE + 1U];
 
-int
-main(int argc, char **argv)
+int main(int argc, char **argv)
 {
 #if PC_HOSTED == 1
 	platform_init(argc, argv);
 #else
-	(void) argc;
-	(void) argv;
+	(void)argc;
+	(void)argv;
 	platform_init();
 #endif
 
 	while (true) {
 		volatile struct exception e;
-		TRY_CATCH(e, EXCEPTION_ALL) {
-			gdb_main();
+		TRY_CATCH (e, EXCEPTION_ALL) {
+			SET_IDLE_STATE(0);
+			while (gdb_target_running && cur_target) {
+				gdb_poll_target();
+
+				// Check again, as `gdb_poll_target()` may
+				// alter these variables.
+				if (!gdb_target_running || !cur_target) {
+					break;
+				}
+				char c = (char)gdb_if_getchar_to(0);
+				if (c == '\x03' || c == '\x04') {
+					target_halt_request(cur_target);
+				}
+				platform_pace_poll();
+#ifdef ENABLE_RTT
+				if (rtt_enabled)
+					poll_rtt(cur_target);
+#endif
+			}
+
+			SET_IDLE_STATE(1);
+			size_t size = gdb_getpacket(pbuf, BUF_SIZE);
+			// If port closed and target detached, stay idle
+			if ((pbuf[0] != 0x04) || cur_target) {
```

Comment:
The braces appear necessary here, because on hosted `SET_IDLE_STATE(x)` is defined as an empty macro.

If you like, I can redefine `SET_IDLE_STATE(x)` to be `do {} while(0)` on hosted, which would get around that error...

Reply:
I went ahead and added that in c86e7964304845fc062749b94b3d3e9bc3ba0467 which fixes this on `hosted`.

Reply:
ah, right.. that mess - yes, good call to turn that into a `do {} while(0)`